app/pipes.go

```diff
@@ -0,0 +1,194 @@
+package app
+
+import (
+	"io"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+
+	"github.com/weaveworks/scope/common/mtime"
+	"github.com/weaveworks/scope/xfer"
+)
+
+const (
+	gcInterval  = 30 * time.Second // we check all the pipes every 30s
+	pipeTimeout = 1 * time.Minute  // pipes are closed when a client hasn't been connected for 1 minute
+	gcTimeout   = 1 * time.Minute  // after another 1 minute, tombstoned pipes are forgotten
+)
+
+// PipeRouter connects incoming and outgoing pipes.
+type PipeRouter struct {
+	sync.Mutex
+	wait  sync.WaitGroup
+	quit  chan struct{}
+	pipes map[string]*pipe
+}
+
+// for each end of the pipe, we keep a reference count & lastUsedTIme,
+// such that we can timeout pipes when either end is inactive.
+type end struct {
+	refCount     int
+	lastUsedTime time.Time
+}
+
+type pipe struct {
+	ui, probe     end
+	tombstoneTime time.Time
+
+	xfer.Pipe
+}
+
+// RegisterPipeRoutes registers the pipe routes
+func RegisterPipeRoutes(router *mux.Router) *PipeRouter {
+	pipeRouter := &PipeRouter{
+		quit:  make(chan struct{}),
+		pipes: map[string]*pipe{},
+	}
+	pipeRouter.wait.Add(1)
+	go pipeRouter.gcLoop()
+	router.Methods("GET").Path("/api/pipe/{pipeID}").HandlerFunc(pipeRouter.handleWs(func(p *pipe) (*end, io.ReadWriter) {
+		uiEnd, _ := p.Ends()
+		return &p.ui, uiEnd
+	}))
+	router.Methods("GET").Path("/api/pipe/{pipeID}/probe").HandlerFunc(pipeRouter.handleWs(func(p *pipe) (*end, io.ReadWriter) {
+		_, probeEnd := p.Ends()
+		return &p.probe, probeEnd
+	}))
+	router.Methods("DELETE").Path("/api/pipe/{pipeID}").HandlerFunc(pipeRouter.deletePipe)
+	router.Methods("POST").Path("/api/pipe/{pipeID}").HandlerFunc(pipeRouter.deletePipe)
+	return pipeRouter
+}
+
+// Stop stops the pipeRouter
+func (pr *PipeRouter) Stop() {
+	close(pr.quit)
+	pr.wait.Wait()
+}
+
+func (pr *PipeRouter) gcLoop() {
+	defer pr.wait.Done()
+	ticker := time.Tick(gcInterval)
+	for {
+		select {
+		case <-pr.quit:
+			return
+		case <-ticker:
+		}
+
+		pr.timeoutPipes()
+		pr.gcPipes()
+	}
+}
+
+func (pr *PipeRouter) timeoutPipes() {
+	pr.Lock()
+	defer pr.Unlock()
+	now := mtime.Now()
+	for id, pipe := range pr.pipes {
+		if pipe.Closed() || (pipe.ui.refCount > 0 && pipe.probe.refCount > 0) {
+			continue
+		}
+
+		if (pipe.ui.refCount == 0 && now.Sub(pipe.ui.lastUsedTime) >= pipeTimeout) ||
+			(pipe.probe.refCount == 0 && now.Sub(pipe.probe.lastUsedTime) >= pipeTimeout) {
+			log.Printf("Timing out pipe %s", id)
+			pipe.Close()
+			pipe.tombstoneTime = now
+		}
+	}
+}
+
+func (pr *PipeRouter) gcPipes() {
```

Comment:
What would a `PipeRouter` GC, if not pipes? (Same for timeout)

Edit: And a lot of the other methods in here.